lab11.md

```diff
@@ -0,0 +1,173 @@
+Before starting this lab create new folder
+```
+mkdir -p lab11 \
+   && cd lab11
+```
+
+## What is HMAC
+
+Currently our functions can get information from everywhere without any restriction. If we want to validate the source of information we can use _Hash-based Message Authentication Code (HMAC)_.
+
+From [alexellis/hmac](https://github.com/alexellis/hmac):
+> HMAC uses a symmetric key that both sender/receiver share ahead of time. The sender will generate a hash when wanting to transmit a message - this data is sent along with the payload. The recipient will then sign payload with the shared key and if the hash matches then the payload is assumed to be from the sender.
+
+This way we close our functions from being invoked with invalid or even dangerous information.
+
+## Using HMAC
+
+For the purpose we will create new python3 function and let's give it a name `hmac-protected`:
+
+```
+$ faas-cli new --lang python3 hmac-protected --prefix="<your-docker-username>"
+```
+
+Add `payload-secret` which will serve as the key that will decrypt the HMAC payload. 
+
+Our `hmac-protected.yml` should look like:
+
+```
+provider:
+  name: faas
+  gateway: http://127.0.0.1:8080
+
+functions:
+  hmac-protected:
+    lang: python3
+    handler: ./hmac-protected
+    image: <your-docker-username>/hmac-protected:latest
+    secrets:
+      - payload-secret
+```
+
+Now populate the `payload-secret` like we did in [lab10](https://github.com/openfaas/workshop/blob/master/lab10.md):
+
+```
+$ echo -n "<your-secret>" | docker secret create payload-secret -
+```
+
+> Note: Remember the string you put in place of  `<your-secret>`
+
+Replace the content of the `handler.py` with the following code:
+
+```
+import os, hmac, hashlib
+
+def validateHMAC(message,secret,hash):
+
+    # The hash starts with the encryption method "sha1=" we need to remove it
+    cleanHash = hash[5:]
+    encodedSecret = secret.encode()
+    encodedMessage = message.encode()
+    expectedMac = hmac.new(encodedSecret,encodedMessage,hashlib.sha1)
+
+    hmacDigest = expectedMac.hexdigest()
+
+    # Comparing the hexidecimal MAC
+    if hmacDigest == cleanHash:
+        return True
+    return False
+
+def handle(req):
+    # We send secret through HMAC header
+    encryptedMessage = os.getenv("Http_Hmac")
+
+    # Read secret from inside the container
+    with open("/var/openfaas/secrets/payload-secret","r") as secretContent:
+        payloadSecret = secretContent.read()
+
+    # Function to validate the HMAC
+    if validateHMAC(req,payloadSecret,encryptedMessage):
+        return "Successfully validated: " + req
+    return "HMAC validation failed."
+```
+
+> The source code is also available at [hmac-protected/hmac-protected/handler.py](./hmac-protected/hmac-protected/handler.py)
+
+* Build, push and deploy the function:
+```
+$ faas up -f ./hmac-protected.yml
+```
+
+### Invoke function
+
+We can sign the header with the `--sign` flag that the faas-cli provides and invoke the function. 
```

Comment:
Let’s add a sentence why we need to sign, pointing to the code lines that require it.